packages/test/test-utils/src/testObjectProvider.ts

```diff
@@ -474,10 +591,339 @@ export class TestObjectProvider implements ITestObjectProvider {
 		return waitContainerToCatchUp_original(container);
 	}
 
-	updateDocumentId(resolvedUrl: IResolvedUrl | undefined) {
+	/**
+	 * {@inheritDoc ITestObjectProvider.updateDocumentId}
+	 */
+	public updateDocumentId(resolvedUrl: IResolvedUrl | undefined) {
 		this._documentIdStrategy.update(resolvedUrl);
 	}
 
+	/**
+	 * {@inheritDoc ITestObjectProvider.resetLoaderContainerTracker}
+	 */
+	public resetLoaderContainerTracker(syncSummarizerClients: boolean = false) {
+		this._loaderContainerTracker.reset();
+		this._loaderContainerTracker = new LoaderContainerTracker(syncSummarizerClients);
+	}
+}
+
+/**
+ * Implements {@link ITestObjectProvider}, but uses different versions to create and load containers.
+ *
+ * @internal
+ */
+export class TestObjectProviderWithVersionedLoad implements ITestObjectProvider {
+	/**
+	 * {@inheritDoc ITestObjectProvider."type"}
+	 */
+	public readonly type = "TestObjectProviderWithVersionedLoad";
+	private _loaderContainerTracker = new LoaderContainerTracker();
+	private _logger: EventAndErrorTrackingLogger | undefined;
+	private readonly _documentIdStrategy: IDocumentIdStrategy;
+	private _documentServiceFactory: IDocumentServiceFactory | undefined;
+	private _urlResolver: IUrlResolver | undefined;
+	// Since documentId doesn't change we can only create/make one container. Call the load functions instead.
+	private _documentCreated = false;
+	// `_loadCount` is used to alternate which version we load the next container with.
+	// If loadCount is even then we will load with the "create" version, and if odd we load with the "load" version.
+	// After each test we will reset loadCount to 0 to ensure we always create the first container with the create version.
+	// Note: This will only affect tests that load a container more than two times.
+	private _loadCount: number = 0;
+
+	constructor(
+		private readonly LoaderConstructorForCreating: typeof Loader,
+		private readonly LoaderConstructorForLoading: typeof Loader,
+		private readonly driverForCreating: ITestDriver,
+		private readonly driverForLoading: ITestDriver,
+		private readonly createFluidEntryPointForCreating: (
+			testContainerConfig?: ITestContainerConfig,
+		) => fluidEntryPoint,
+		private readonly createFluidEntryPointForLoading: (
+			testContainerConfig?: ITestContainerConfig,
+		) => fluidEntryPoint,
+	) {
+		this._documentIdStrategy = getDocumentIdStrategy(driverForCreating.type);
+	}
+
+	/**
+	 * {@inheritDoc ITestObjectProvider.logger}
+	 */
+	public get logger(): EventAndErrorTrackingLogger {
+		if (this._logger === undefined) {
+			this._logger = new EventAndErrorTrackingLogger(
```

Comment:
I'm guessing some of this is just copied from TestObjectProvider, but any particular reason for delaying creation of the logger until access?  If not, consider initializing in the constructor so it can always be defined and `readonly`.  And can drop the get/set

If there's a reason for delay-creation, you could consider using ??= syntax here, and you probably don't need the setter since it's private and doesn't do anything special.